File: Blind150/1.Array&Hash/GO/valid_anagram.go
// https://leetcode.com/problems/valid-anagram/description/


// Why Use rune(s[i]) in the isAnagram Example?
// Unicode Characters: If the string contains non-ASCII characters (e.g., characters with accents, Chinese characters), they may consist of multiple bytes. s[i] would only give you one byte, which might not represent a complete character. rune(s[i]) ensures you're dealing with the full Unicode code point.
// Consistency: By using rune(s[i]), you're ensuring that your code correctly handles all characters, not just single-byte ASCII characters.


package main

import (
	"fmt"
)

func isAnagram(s, t string) bool {

	if len(s) != len(t){
		return false
	}

	countS := make(map[rune]int)
	countT := make(map[rune]int)

	for i:=0 ; i<len(s); i++ {

		countS[rune(s[i])] ++
		countT[rune(t[i])] ++

	}

	for key, val := range countS{
		if countT[key] != val{
			return false
		}
	}


	return true
}

File: Blind150/1.Array&Hash/Python/valid_anagram.py
# https://leetcode.com/problems/valid-anagram/description/


from collections import Counter
class Solution1:
    def isAnagram(self, s: str, t: str) -> bool:

        return Counter(s) == Counter(t)




######### OR WE CAN WRITE LIKE THIS

class Solution2:
    def isAnagram(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False


        counts, countt = {}, {}

        for cs, ct in zip(s,t):

            counts[cs] = 1 + counts.get(cs,0)
            countt[ct] = 1 + countt.get(ct,0)

        return counts == countt
